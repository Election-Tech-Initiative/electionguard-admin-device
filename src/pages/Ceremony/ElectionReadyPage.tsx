import React, { useContext, useState, useEffect, useCallback } from 'react'
import { withRouter, RouteComponentProps } from 'react-router-dom'
import Main, { MainChild } from '../../components/Main'
import Sidebar from '../../components/Sidebar'
import AdminContext from '../../contexts/adminContext'
import Screen from '../../components/Screen'
import LinkButton from '../../components/LinkButton'
import Prose from '../../components/Prose'
import { ElectionGuardStatus } from '../../electionguard'
import SidebarFooter from '../../components/SidebarFooter'
import LoadAdminDrive from '../LoadAdminDrive'
import UsbContext from '../../contexts/usbContext'
import {
  adminDriveIndex,
  electionFile,
  configFile,
  mapFile,
} from '../../components/UsbManager'

const ElectionReadyPage = (props: RouteComponentProps) => {
  const {
    election,
    electionMap,
    electionGuardConfig,
    setElectionGuardStatus,
  } = useContext(AdminContext)
  const { adminDriveConnected, connect, disconnect, write } = useContext(
    UsbContext
  )
  const [savingElectionData, setSavingElectionData] = useState(false)

  const startMonitoringDrives = useCallback(connect, [])
  const stopMonitoringDrives = useCallback(disconnect, [])
  useEffect(() => {
    startMonitoringDrives()
    return () => stopMonitoringDrives()
  }, [startMonitoringDrives, stopMonitoringDrives])

  const completeElectionSetup = () => {
    setSavingElectionData(true)
  }

  if (savingElectionData && !adminDriveConnected) {
    return <LoadAdminDrive />
  }

  if (savingElectionData && adminDriveConnected) {
    const writeFilesToAdminDrive = async () => {
      let result = await write(adminDriveIndex, electionFile, election)
      if (!result.success) {
        throw new Error('failed to write election data')
      }

      result = await write(adminDriveIndex, configFile, {
        ...electionGuardConfig,
        status: ElectionGuardStatus.TallyVotes,
      })
      if (!result.success) {
        throw new Error('failed to write election configuration data')
      }

      result = await write(adminDriveIndex, mapFile, electionMap)
      if (!result.success) {
        throw new Error('failed to write election map data')
      }
    }

    const { history } = props
    writeFilesToAdminDrive().then(() => {
      setElectionGuardStatus(ElectionGuardStatus.TallyVotes)
      history.push('/start')
    })
  }

  return (
    <Screen>
      <Main>
        <MainChild center>
          <Prose textCenter>
            <h1>Election Ready</h1>
            <hr />
            <p>Election setup is complete. </p>
          </Prose>
        </MainChild>
      </Main>
      <Sidebar footer={<SidebarFooter />}>
        <p>
          <LinkButton
            big
            primary={!savingElectionData}
            disabled={savingElectionData}
            onPress={completeElectionSetup}
            id="election-complete"
            aria-label="Election Complete"
          >
            Complete
          </LinkButton>
        </p>
      </Sidebar>
    </Screen>
  )
}

export default withRouter(ElectionReadyPage)
