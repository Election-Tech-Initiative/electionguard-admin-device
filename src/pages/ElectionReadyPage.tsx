import React, { useContext } from 'react'
import styled from 'styled-components'
import Main, { MainChild } from '../components/Main'
import Sidebar from '../components/Sidebar'
import ElectionContext from '../contexts/electionContext'
import ElectionInfo from '../components/ElectionInfo'
import Screen from '../components/Screen'
import LinkButton from '../components/LinkButton'
import Prose from '../components/Prose'
import { ElectionGuardStatus } from '../config/types'

const LogoImage = styled.img`
  display: flex;
  margin: 0 auto;
  max-width: 12rem;
`

const ElectionReadyPage = () => {
  const { election, electionGuardStatus, setElectionGuardStatus } = useContext(
    ElectionContext
  )
  if (electionGuardStatus === ElectionGuardStatus.KeyCeremony) {
    setElectionGuardStatus(ElectionGuardStatus.TallyVotes)
  }

  return (
    <Screen>
      <Main>
        <MainChild center>
          <Prose textCenter>
            <h1>Election Ready</h1>
            <hr />
            <p>Election setup is complete. </p>
          </Prose>
        </MainChild>
      </Main>
      <Sidebar
        footer={
          <>
            <hr />
            <ElectionInfo election={election} precinctId="" horizontal />
            <hr />
            <LogoImage
              alt="Election Guard Logo"
              src="/images/electionguard.svg"
            />
          </>
        }
      >
        <p>
          <LinkButton
            big
            primary
            to="/start"
            id="election-complete"
            aria-label="Election Complete"
          >
            Complete
          </LinkButton>
        </p>
      </Sidebar>
    </Screen>
  )
}

export default ElectionReadyPage
