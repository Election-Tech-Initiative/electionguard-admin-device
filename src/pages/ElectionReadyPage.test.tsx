import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import ElectionReadyPage from './ElectionReadyPage'

it('renders ElectionReadyPage', async () => {
  const { container } = render(
    <Route path="/" component={ElectionReadyPage} />,
    {
      route: '/',
    }
  )
  expect(container.firstChild).toMatchSnapshot()
})
