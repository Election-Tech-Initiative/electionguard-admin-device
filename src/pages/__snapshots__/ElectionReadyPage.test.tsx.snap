// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ElectionReadyPage 1`] = `
<main
  class="sc-htpNat eUEnuv"
>
  <div
    class="sc-bxivhb fEFIwD"
  >
    <div
      class="sc-ifAKCX cObZIz"
    >
      <h1>
        Election Ready
      </h1>
      <p>
        Election setup is complete. Please return to main menu.
      </p>
      <button
        aria-label="Return to Start"
        class="sc-bwzfXH kHZqpR"
        id="return-to-start"
        role="option"
        type="button"
      >
        Return
      </button>
    </div>
  </div>
</main>
`;
